Make artifact and source cache query explicit

juergbi opened the pull request:
Artifact and source cache query can be fairly expensive as it checks the presence of all artifact and source blobs, respectively. This will be especially true with the upcoming optional support for remote caching (reducing blob downloads with remote execution to a minimum) in PR #1431.

This branch makes cache queries more explicit in the code and puts the `Stream` class in control in preparation for parallelization of cache queries.

This is a simplified replacement for PR #1430, not yet moving the cache queries to scheduled jobs, thus avoiding a performance hit in certain use cases.

juergbi committed:
tests/frontend/push.py: Allow pushing of dependencies

The assertions in `test_push_after_pull` are too strict. Pushing
dependencies to the second (empty) artifact server should not cause a
test failure.

juergbi committed:
_artifact.py: Make cache query explicit

Cache query can be fairly expensive as it checks the presence of all
artifact blobs. Make this more explicit with a `query_cache()` method,
instead of implicitly querying the cache on the first call to
`cached()`.

juergbi committed:
_pipeline.py: Drop the optimization for cached elements in the planner

The overhead of planning already cached elements and unneeded build-only
dependencies should be fairly small as unneeded jobs can still be
skipped. This optimization was also already disabled for non-strict
build plans with a remote artifact cache.

This change is necessary in preparation for parallelizing cache queries.

juergbi committed:
_elementsources.py: Make cache query explicit

Cache query can be fairly expensive as it checks the presence of all
blobs. Make this more explicit with a `query_cache()` method, instead of
implicitly querying the cache on the first call to `cached()`.

juergbi committed:
fetchqueue.py: Don't skip elements with a cached failure

The build queue requires the sources to be available for all elements
where `_cached_success()` returns `False`. This includes elements with a
cached failure.

juergbi committed:
Call _initialize_state() in Element._new_from_load_element()

With the cache queries moved out, `_initialize_state()` is fairly
lightweight and can be called earlier.

juergbi committed:
element.py: Combine cache query and pull into `_load_artifact()`

In non-strict mode cache query and pull are intertwined as we prefer
pulling the strict artifact to a cache lookup with the weak cache key.

This replaces `__update_artifact_state()` and `_pull()` with a combined
`_load_artifact()` method, which supports cache query with deferred
pulling. This provides correct behavior with the flexibility of split or
combined cache query and pull.

juergbi committed:
Move artifact and source cache query trigger to Stream class

This is in preparation for parallelization of cache queries.

gtristan commented on src/buildstream/_artifact.py:
```diff
@@ -550,11 +550,7 @@ def get_dependency_artifact_names(self):
-        if self._cached is not None:
-            return self._cached
-
+    def query_cache(self):
```
This is rather orthogonal to the patch, but I think we need to cleanup how this function behaves differently depending on global values decided in the `Context`.

I don't feel like it is safe to have such a deep function call behaving differently *whole sale* based on user configuration, and think rather we should be passing these values in as arguments when checking what *"domains"* of an artifact are locally cached (i.e. is only the ref cached, are the files available locally, etc).

* `Context.require_artifact_files`

  This is only ever set to `False` by `Stream` in the case that remote execution is enabled

* `Context.require_artifact_directories`

  This is unconditionally `True` all the time, maybe this historically meant something, but we forgot to remove this boolean once things changed such that it is always true (at least grep seems to tell me nothing ever sets this value to `False`)

* `Element._artifact_files_required()`

  Even here I feel like it may happen that an `Element` which does evaluate to `False` for this function (i.e. it *"requires artifact files"*) could justifiably decide to ask the `Artifact` whether the files are cached or not

We can resolve this separately from the branch, but as I am reading it this jumps out at me and strikes me as problematic so just raising it here.

juergbi replied:
PR #1431 completely drops the two variables and the method you mention above, see commit 87e93c1fc6a45c83823a541719f24e1aa8f92f56.

gtristan replied:
Sounds great !

gtristan commented on src/buildstream/_artifactelement.py:
```diff
@@ -131,9 +131,9 @@ def clear_artifact_name_cache(cls):
     #
-    def _pull_done(self):
-        super()._pull_done()
+    def _load_artifact_done(self):
```
I think we need to streamline how `ArtifactElement` instances are initialized with this.

This branch changes a bit the semantics of how an artifact is loaded onto an element, but does not appear to update `Element.__initialize_from_artifact()`, perhaps we can have `Element.__initialize_from_artifact()` directly call `Element._load_artifact()` or such.

Another point, is that loading the artifact from the local cache is obviously a hard requirement when loading any `ArtifactElement`, regardless of whether `Stream.query_cache()` was called on the element.

I think this branch does not exactly change the current initialization (and deferring the artifact loading is planned in a later branch), in any case, it should be clear somehow (perhaps in the `Stream.query_cache()` API comments) that this will end up being a `noop` for `ArtifactElements` which are anyway synchronously loaded at instantiation time.

juergbi replied:
Using the `_load_artifact()` code path for artifact element early loading seems sensible. I've pushed a separate commit for this.

In `Stream.query_cache()` I've added an explicit no-op case if the artifact has already been loaded.

gtristan replied:
Thanks :)

My concern was mostly to ensure that the state of an artifact loaded early on, or an artifact loaded as a result of pulling it, would be as close as possible.

gtristan replied:
Looking at your patch, it seems the code has changed more than I expected it might have... anyway, looks good to me :)

juergbi committed:
Use _load_artifact() for ArtifactElement

Artifact elements require access to the artifact early on to walk
dependencies. Use `_load_artifact()` for this instead of a custom code
path.
